### PR TITLE
[QUIC] Update to msquic 2

### LINKED ISSUE #64554: [QUIC] Update to msquic 2.0
Once MsQuic 2.0 is released: https://github.com/microsoft/msquic/pull/2254, we should update to it.
Note that there are breaking changes in 2.0.
Also we'll need to update published packages to consume in tests/stress on Linux (I'm not sure, whether we build it from source or download from packages.microsoft.com atm).

### PR DESCRIPTION
**Pull request by ManickaP**
Fixes https://github.com/dotnet/runtime/issues/64554, https://github.com/dotnet/runtime/issues/53090

Please review, but don't merge it yet, I need to update our testing docker images first.

### CODE REVIEW

File: src/libraries/System.Net.Quic/src/System/Net/Quic/Implementations/MsQuic/MsQuicConnection.cs
Comment (ManickaP) [inline src/libraries/System.Net.Quic/src/System/Net/Quic/Implementations/MsQuic/MsQuicConnection.cs:697]: We still get SHUTDOWN_COMPLETE if the connection start fails. If we free the context object here, the callback gets random memory and blows up.

File: src/libraries/System.Net.Quic/src/System/Net/Quic/Implementations/MsQuic/MsQuicListener.cs
Comment (ManickaP) [inline src/libraries/System.Net.Quic/src/System/Net/Quic/Implementations/MsQuic/MsQuicListener.cs:162]: This is absolutely terrible solution and we need to come up with better one. I'm taking this into account for Listener API.

File: src/libraries/System.Net.Quic/tests/FunctionalTests/MsQuicTests.cs
Comment (ManickaP) [inline src/libraries/System.Net.Quic/tests/FunctionalTests/MsQuicTests.cs:389]: We need to wait for stream start callback event in all cases.

**Comment by ManickaP** [comment]
@wfurt looking at this, we should already be consuming latest and greatest msquic:
https://github.com/dotnet/dotnet-buildtools-prereqs-docker/blob/dcc36f08baa367d98f9c6cfcdde28fc0d5f6c6d9/src/fedora/34/helix/amd64/Dockerfile#L25-L34

How come it didn't broke the tests? Or is it because we need to update the image tag in our test pipeline and we haven't done it yet?

File: src/libraries/System.Net.Quic/src/System/Net/Quic/Implementations/MsQuic/MsQuicConnection.cs
Comment (CarnaViire) [reply src/libraries/System.Net.Quic/src/System/Net/Quic/Implementations/MsQuic/MsQuicConnection.cs:697]: Can we please leave this as a comment in the code?

File: src/libraries/System.Net.Quic/src/System/Net/Quic/Implementations/MsQuic/MsQuicListener.cs
Comment (CarnaViire) [reply src/libraries/System.Net.Quic/src/System/Net/Quic/Implementations/MsQuic/MsQuicListener.cs:162]: And this too, can we leave a TODO comment?.. Just so it will not slip from our attention

File: src/libraries/System.Net.Quic/tests/FunctionalTests/MsQuicTests.cs
Comment (rzikm) [inline src/libraries/System.Net.Quic/tests/FunctionalTests/MsQuicTests.cs:310]: Is this change intended?

File: src/libraries/System.Net.Quic/src/System/Net/Quic/Implementations/MsQuic/MsQuicListener.cs
Comment (CarnaViire) [inline src/libraries/System.Net.Quic/src/System/Net/Quic/Implementations/MsQuic/MsQuicListener.cs:215]: StopAsync?..

File: src/libraries/System.Net.Quic/src/System/Net/Quic/Implementations/MsQuic/MsQuicStream.cs
Comment (rzikm) [inline src/libraries/System.Net.Quic/src/System/Net/Quic/Implementations/MsQuic/MsQuicStream.cs:181]: Since the stream start is asynchronous now, do we need to do anything to prevent stream operations before we get the StartComplete callback?

**Review by rzikm — APPROVED** [review]
LGTM, just few questions

File: src/libraries/System.Net.Quic/tests/FunctionalTests/MsQuicTests.cs
Comment (ManickaP) [reply src/libraries/System.Net.Quic/tests/FunctionalTests/MsQuicTests.cs:310]: Yes, `Conditional` makes the test ignore the `SkipException`, otherwise it fails.
I assume the tests in CI might run without IPv6 support (only docker) so it didn't cause noise on your PR, but it was on my machine.

File: src/libraries/System.Net.Quic/src/System/Net/Quic/Implementations/MsQuic/MsQuicStream.cs
Comment (CarnaViire) [reply src/libraries/System.Net.Quic/src/System/Net/Quic/Implementations/MsQuic/MsQuicStream.cs:181]: Most possibly, but I believe it will be part of https://github.com/dotnet/runtime/issues/67302 (I also mentioned it in additional thoughts there)

File: src/libraries/System.Net.Quic/tests/FunctionalTests/MsQuicTests.cs
Comment (rzikm) [reply src/libraries/System.Net.Quic/tests/FunctionalTests/MsQuicTests.cs:310]: I was wondering why `SkipException` sometimes fails and other times skips...., thanks.

File: src/libraries/System.Net.Quic/src/System/Net/Quic/Implementations/MsQuic/MsQuicStream.cs
Comment (CarnaViire) [inline src/libraries/System.Net.Quic/src/System/Net/Quic/Implementations/MsQuic/MsQuicStream.cs:43]: Is it intentional that it will be "success" even before the event arrives?

File: src/libraries/System.Net.Quic/src/System/Net/Quic/Implementations/MsQuic/MsQuicStream.cs
Comment (CarnaViire) [reply src/libraries/System.Net.Quic/src/System/Net/Quic/Implementations/MsQuic/MsQuicStream.cs:181]: > behavior depending on max stream limit is broken now

only throwing on reaching the limit is broken, right? the wait loop implemented in H3 should be fine? @ManickaP

File: src/libraries/System.Net.Quic/src/System/Net/Quic/Implementations/MsQuic/MsQuicConnection.cs
Comment (ManickaP) [reply src/libraries/System.Net.Quic/src/System/Net/Quic/Implementations/MsQuic/MsQuicConnection.cs:697]: I don't think it makes sense now since we only free the GC handle in failed ctors and in SHUTDOWN_COMPLETE, which makes sense.
Before this change, the call to free in failed connect was rather out-of-place.

File: src/libraries/System.Net.Quic/src/System/Net/Quic/Implementations/MsQuic/MsQuicStream.cs
Comment (ManickaP) [reply src/libraries/System.Net.Quic/src/System/Net/Quic/Implementations/MsQuic/MsQuicStream.cs:181]: That's my understanding.

File: src/libraries/System.Net.Quic/src/System/Net/Quic/Implementations/MsQuic/MsQuicStream.cs
Comment (ManickaP) [reply src/libraries/System.Net.Quic/src/System/Net/Quic/Implementations/MsQuic/MsQuicStream.cs:43]: Yes, we check it only in shutdown complete to unblock pending operations with proper exception. There's no point in distinguishing between start not called yet and  called and succeeded.
I also assume this will get reworked with #67302.

File: src/libraries/System.Net.Quic/src/System/Net/Quic/Implementations/MsQuic/MsQuicListener.cs
Comment (ManickaP) [reply src/libraries/System.Net.Quic/src/System/Net/Quic/Implementations/MsQuic/MsQuicListener.cs:162]: Done

File: src/libraries/System.Net.Quic/src/System/Net/Quic/Implementations/MsQuic/MsQuicListener.cs
Comment (ManickaP) [reply src/libraries/System.Net.Quic/src/System/Net/Quic/Implementations/MsQuic/MsQuicListener.cs]: Done

**Review by CarnaViire — APPROVED** [review]
LGTM, thanks!

File: src/libraries/System.Net.Quic/src/System/Net/Quic/Implementations/MsQuic/Interop/MsQuicEnums.cs
Comment (nibanks) [inline src/libraries/System.Net.Quic/src/System/Net/Quic/Implementations/MsQuic/Interop/MsQuicEnums.cs:72]: We should look into using MsQuic's [generated interop](https://github.com/microsoft/msquic/blob/main/src/cs/lib/msquic_generated.cs) layer instead of doing all this manually in .NET. Several of these enums are out of date, and don't include all the latest values.

File: src/libraries/System.Net.Quic/src/System/Net/Quic/Implementations/MsQuic/MsQuicListener.cs
Comment (nibanks) [inline src/libraries/System.Net.Quic/src/System/Net/Quic/Implementations/MsQuic/MsQuicListener.cs:163]: We added this async stop functionality at the request of .NET. 😄 We were told it would make things easier for you!

File: src/libraries/System.Net.Quic/src/System/Net/Quic/Implementations/MsQuic/MsQuicStream.cs
Comment (nibanks) [reply src/libraries/System.Net.Quic/src/System/Net/Quic/Implementations/MsQuic/MsQuicStream.cs:181]: FYI, it might be worth having a conversation about the `FAIL_BLOCKED` usage here. MsQuic allows you to start a stream and send on it before starting a connection (practically a requirement for 0-RTT scenarios), and without this flag it allows you to queue safely. But with this flag, it would just fail because you are blocked on starting the connection. It _might_ make sense to use this in your HTTP/3 scenarios, but not universally.

File: src/libraries/System.Net.Quic/src/System/Net/Quic/Implementations/MsQuic/Interop/MsQuicEnums.cs
Comment (nibanks) [reply src/libraries/System.Net.Quic/src/System/Net/Quic/Implementations/MsQuic/Interop/MsQuicEnums.cs:72]: Sounds good!

File: src/libraries/System.Net.Quic/src/System/Net/Quic/Implementations/MsQuic/MsQuicListener.cs
Comment (ManickaP) [reply src/libraries/System.Net.Quic/src/System/Net/Quic/Implementations/MsQuic/MsQuicListener.cs:163]: Yes, I have it in my TODO list and plan to address. This is just a hack to get it working and be able to move forward, I'm aware it's ugly as heck :smiley:

**Comment by ManickaP** [comment]
Outerloop failure are `badssl.com` is down: #67437 
H/3 stress didn't run due to docker issues me and @rzikm observed locally, we should investigate more.

File: src/libraries/Common/tests/System/Net/Http/Http3LoopbackConnection.cs
Comment (CarnaViire) [inline src/libraries/Common/tests/System/Net/Http/Http3LoopbackConnection.cs:256]: I believe we should bring this back? #58078 is not fixed yet

Comment (ManickaP) [reply src/libraries/Common/tests/System/Net/Http/Http3LoopbackConnection.cs:256]: I'll bring it back, but interestingly I haven't seen it in any of my local runs nor pipelines...

Comment (CarnaViire) [reply src/libraries/Common/tests/System/Net/Http/Http3LoopbackConnection.cs:256]: It's a race that is not so easy to catch I guess 😅 but it definitely exists

File: src/libraries/System.Net.Quic/src/System/Net/Quic/Implementations/MsQuic/MsQuicStream.cs
Comment (ManickaP) [reply src/libraries/System.Net.Quic/src/System/Net/Quic/Implementations/MsQuic/MsQuicStream.cs:181]: Yes, we do plan to rework stream starting in #67302. We have more issues around this than just 0-RTT :smile:

**Review by CarnaViire — APPROVED** [review]
LGTM, thanks!

**Comment by ManickaP** [comment]
Failing tests seem like fallout from #67037, I'll revert the change around `EnableReceive` and we'll see it helps.
My theory is that at that time of #67037 we already had updated msquic in the transport package for Windows so the CI didn't run the tests there (IsSupported would fail due to version incompatibility) and this is actually the first time they are run.

If my theory proves true, @rzikm you might want to revisit this.

**Pull request merged**
